syntax = "proto3";

import "google/protobuf/timestamp.proto";

message FsRequest
{
    int64 serial = 1;   // 序号
    optional string path = 2;
}

message FsResponse
{
    int64 serial = 1;   // 序号
    bool accepted = 2;
    uint32 port = 3;
}

message FsSendFileRequest
{
    int64 serial = 1;   // 序号
    string path = 2;    // 文件名
}

message FsSendFileResponse
{
    int64 serial = 1;   // 序号
    bool accepted = 2;
}

message FsStat
{
    uint64 ino = 1;
    uint32 mode = 2;    // file type and mode
    uint64 nlink = 3;
    uint32 uid = 4;
    uint32 gid = 5;
    uint64 size = 6;    // total size, in bytes
    google.protobuf.Timestamp atime = 7;    // last access of file data
    google.protobuf.Timestamp mtime = 8;    // last modification of file data
    google.protobuf.Timestamp ctime = 9;    // last status change timestamp (time of last change to the inode)
}

message FsFileInfo
{
    int32 flags = 1;
    bool writepage = 2;
    bool direct_io = 3;
    bool keep_cache = 4;
    bool flush = 5;
    bool nonseekable = 6;
    bool cache_readdir = 7;
    uint64 fh = 8;
    uint64 lock_owner = 9;
    uint32 poll_events = 10;
    bool noflush = 11;
}

message FsMethodGetAttrRequest
{
    int64 serial = 1;   // 序号
    string path = 2;
    optional FsFileInfo fi = 3;
}

message FsMethodGetAttrResponse
{
    int64 serial = 1;   // 序号
    int32 result = 2;
    FsStat stat = 3;
}

message FsMethodReadDirRequest
{
    int64 serial = 1;   // 序号
    string path = 2;
    uint64 offset = 3;
    optional FsFileInfo fi = 4;
}

message FsMethodReadDirResponse
{
    int64 serial = 1;   // 序号
    int32 result = 2;
    repeated string item = 3;
}

message FsMethodReadRequest
{
    int64 serial = 1;   // 序号
    uint64 offset = 2;
    uint64 size = 3;    // max 2GB
    optional FsFileInfo fi = 4;
}

message FsMethodReadResponse
{
    int64 serial = 1;   // 序号
    int32 result = 2;
    bytes data = 3;
}

message FsMethodOpenRequest
{
    int64 serial = 1;   // 序号
    string path = 2;
    optional FsFileInfo fi = 3;
}

message FsMethodOpenResponse
{
    int64 serial = 1;   // 序号
    int32 result = 2;
    optional uint64 fh = 3;
}

message FsMethodReleaseRequest
{
    int64 serial = 1;   // 序号
    string path = 2;
    optional FsFileInfo fi = 3;
}

message FsMethodReleaseResponse
{
    int64 serial = 1;   // 序号
}
